$NetBSD: patch-aa,v 1.15 2015/01/25 16:13:13 taca Exp $

* Relax to allow some attribute could be nil.
* Complete "http://" to homepage attribute if it isn't begin with it.
* Accpet date with Time, but ignore time part.
* Skip checking gem.build_complete existence of file.
* Allow duplicate dependencies but warn them.

--- lib/rubygems/specification.rb.orig	2014-11-30 04:25:42.000000000 +0000
+++ lib/rubygems/specification.rb
@@ -172,6 +172,12 @@ class Gem::Specification < Gem::BasicSpe
     @@default_value[k].nil?
   }
 
+  # under pkgsrc environment
+  @@pkgsrc_nil_attributes = {}
+  [:authors, :cert_chain, :executables, :extensions, :extra_rdoc_files, :licenses, :required_rubygems_version, :test_files].each do |k|
+    @@pkgsrc_nil_attributes[k] = true
+  end
+
   ######################################################################
   # :section: Required gemspec attributes
 
@@ -1779,7 +1785,7 @@ class Gem::Specification < Gem::BasicSpe
   # directory.
 
   def gem_build_complete_path # :nodoc:
-    File.join extension_dir, 'gem.build_complete'
+    File.join extension_dir
   end
 
   ##
@@ -2456,8 +2462,25 @@ class Gem::Specification < Gem::BasicSpe
     extend Gem::UserInteraction
     normalize
 
-    nil_attributes = self.class.non_nil_attributes.find_all do |name|
-      instance_variable_get("@#{name}").nil?
+    nil_attributes = self.class.non_nil_attributes.find_all do |attrname|
+      val = instance_variable_get("@#{attrname}")
+      if val.nil?
+        if @@pkgsrc_nil_attributes.key?(attrname)
+          case attrname
+          when :required_rubygems_version
+            val = Gem::Requirement.default
+          else
+            val = []
+          end
+          instance_variable_set("@#{attrname}", val)
+          false
+        end
+      end
+      if attrname == :authors and val.empty?
+        val = ['not specified']
+        self.authors = val
+        false
+      end
     end
 
     unless nil_attributes.empty? then
@@ -2601,10 +2624,15 @@ http://opensource.org/licenses/alphabeti
       raise Gem::InvalidSpecificationException, "#{lazy} is not a summary"
     end
 
-    if homepage and not homepage.empty? and
-       homepage !~ /\A[a-z][a-z\d+.-]*:/i then
-      raise Gem::InvalidSpecificationException,
-            "\"#{homepage}\" is not a URI"
+    if homepage and not homepage.empty?
+      if @homepage =~ /\A[a-z][a-z\d+.-]*/i and
+          @homepage !~ /\A[a-z][a-z\d+.-]*:/i
+        @homepage = "http://" + @homepage
+      end
+      if homepage !~ /\A[a-z][a-z\d+.-]*:/i then
+        raise Gem::InvalidSpecificationException,
+              "\"#{homepage}\" is not a URI"
+      end
     end
 
     # Warnings
@@ -2647,7 +2675,7 @@ http://opensource.org/licenses/alphabeti
 
     dependencies.each do |dep|
       if prev = seen[dep.name] then
-        raise Gem::InvalidSpecificationException, <<-MESSAGE
+        warn <<-MESSAGE
 duplicate dependency on #{dep}, (#{prev.requirement}) use:
     add_runtime_dependency '#{dep.name}', '#{dep.requirement}', '#{prev.requirement}'
         MESSAGE
